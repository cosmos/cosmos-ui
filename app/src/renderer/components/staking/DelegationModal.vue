<template>
  <action-modal
    id="delegation-modal"
    ref="actionModal"
    :submit-fn="submitForm"
    :validate="validateForm"
    title="Delegate"
    class="delegation-modal"
    submission-error-prefix="Delegating failed"
  >
    <tm-form-group
      class="action-modal-form-group"
      field-id="to"
      field-label="To"
    >
      <tm-field id="to" v-model="to" type="text" readonly />
    </tm-form-group>

    <tm-form-group
      v-if="fromOptions.length > 1"
      class="action-modal-form-group"
      field-id="from"
      field-label="From"
    >
      <tm-field
        id="from"
        v-model="selectedIndex"
        :title="from"
        :options="fromOptions"
        type="select"
      />
    </tm-form-group>
    <tm-form-group
      :error="$v.amount.$error && $v.amount.$invalid"
      class="action-modal-form-group"
      field-id="amount"
      field-label="Amount"
    >
      <span class="input-suffix">{{ denom }}</span>
      <tm-field
        id="amount"
        v-model="amount"
        type="number"
        placeholder="Amount"
      />
      <tm-form-msg
        v-if="balance === 0"
        :msg="`doesn't hold any ${denom}s`"
        name="Wallet"
        type="custom"
      />
      <tm-form-msg
        v-else-if="$v.amount.$error && (!$v.amount.required || amount === 0)"
        name="Amount"
        type="required"
      />
      <tm-form-msg
        v-else-if="$v.amount.$error && !$v.amount.between"
        :max="$v.amount.$params.between.max"
        :min="$v.amount.$params.between.min"
        name="Amount"
        type="between"
      />
    </tm-form-group>
  </action-modal>
</template>

<script>
import { mapGetters } from "vuex"
import { required, between, integer } from "vuelidate/lib/validators"
import Modal from "common/TmModal"
import TmBtn from "common/TmBtn"
import TmField from "common/TmField"
import TmFormGroup from "common/TmFormGroup"
import TmFormMsg from "common/TmFormMsg"
import ActionModal from "common/ActionModal"

export default {
  name: `delegation-modal`,
  components: {
    Modal,
    TmBtn,
    TmField,
    TmFormGroup,
    TmFormMsg,
    ActionModal
  },
  props: {
    fromOptions: {
      type: Array,
      required: true
    },
    to: {
      type: String,
      required: true
    },
    validator: {
      type: Object,
      required: true
    },
    denom: {
      type: String,
      required: true
    }
  },
  data: () => ({
    amount: null,
    selectedIndex: 0
  }),
  computed: {
    ...mapGetters([`wallet`, `delegates`]),
    balance() {
      return this.fromOptions[this.selectedIndex].maximum
    },
    from() {
      return this.fromOptions[this.selectedIndex].address
    }
  },
  methods: {
    open() {
      this.$refs.actionModal.open()
    },
    validateForm() {
      this.$v.$touch()

      return !this.$v.$invalid
    },
    async submitDelegation(submitType, password) {
      await this.$store.dispatch(`submitDelegation`, {
        validator_addr: this.validator.operator_address,
        amount: String(this.amount),
        submitType,
        password: password
      })

      this.$store.commit(`notify`, {
        title: `Successful delegation!`,
        body: `You have successfully delegated your ${this.denom}s`
      })
    },
    async submitRedelegation(submitType, password) {
      const validatorSrc = this.delegates.delegates.find(
        v => this.from === v.operator_address
      )
      await this.$store.dispatch(`submitRedelegation`, {
        validatorSrc,
        validatorDst: this.validator,
        amount: String(this.amount),
        submitType,
        password: password
      })

      this.$store.commit(`notify`, {
        title: `Successful redelegation!`,
        body: `You have successfully redelegated your ${this.denom}s`
      })
    },
    async submitForm(submitType, password) {
      if (this.from === this.wallet.address) {
        await this.submitDelegation(submitType, password)
      } else {
        await this.submitRedelegation(submitType, password)
      }
    }
  },
  validations() {
    return {
      amount: {
        required,
        integer,
        between: between(1, this.balance)
      }
    }
  }
}
</script>
