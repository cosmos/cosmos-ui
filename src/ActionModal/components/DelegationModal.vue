<template>
  <ActionModal
    id="delegation-modal"
    ref="actionModal"
    :validate="validateForm"
    :amount="amount"
    title="Stake"
    class="delegation-modal"
    submission-error-prefix="Staking failed"
    :transaction-data="transactionData"
    :notify-message="notifyMessage"
    feature-flag="delegate"
    @close="clear"
    @txIncluded="onSuccess"
  >
    <TmFormGroup class="action-modal-form-group">
      <div class="form-message notice">
        <span>
          It will take 21 days to unlock your tokens after they are staked.
          There is a risk that some tokens will be lost depending on the
          behaviour of the validator you choose.
        </span>
      </div>
    </TmFormGroup>
    <TmFormGroup class="action-modal-form-group" field-id="to" field-label="To">
      <TmField
        id="to"
        v-model="targetValidator.operatorAddress"
        type="text"
        readonly
      />
      <TmFormMsg
        v-if="targetValidator.status === 'INACTIVE'"
        :msg="
          `You are about to stake to an inactive validator (${targetValidator.statusDetailed})`
        "
        type="custom"
        class="tm-form-msg--desc"
      />
    </TmFormGroup>
    <TmFormGroup
      class="action-modal-form-group"
      field-id="from"
      field-label="From"
    >
      <TmField
        id="from"
        v-model="selectedIndex"
        :title="from"
        :options="fromOptions"
        type="select"
      />
    </TmFormGroup>
    <TmFormGroup
      :error="$v.amount.$error && $v.amount.$invalid"
      class="action-modal-form-group"
      field-id="amount"
      field-label="Amount"
    >
      <span class="input-suffix max-button">{{ denom }}</span>
      <TmFieldGroup>
        <TmField
          id="amount"
          v-model="amount"
          v-focus
          placeholder="0"
          class="tm-field-addon"
          type="number"
          @keyup.enter.native="enterPressed"
        />
        <TmBtn
          type="button"
          class="secondary addon-max"
          value="Set Max"
          @click.native="setMaxAmount()"
        />
      </TmFieldGroup>
      <span class="form-message">
        Available to stake:
        {{ maxAmount }}
        {{ denom }}s
      </span>
      <TmFormMsg
        v-if="balance === 0"
        :msg="`doesn't have any ${denom}s`"
        name="Wallet"
        type="custom"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && !$v.amount.decimal"
        name="Amount"
        type="numeric"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && (!$v.amount.required || amount === 0)"
        name="Amount"
        type="required"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && !$v.amount.between"
        :max="$v.amount.$params.between.max"
        :min="$v.amount.$params.between.min"
        name="Amount"
        type="between"
      />
      <TmFormMsg
        v-else-if="isMaxAmount()"
        msg="You are about to use all your tokens for this transaction. Consider leaving a little bit left over to cover the network fees."
        type="custom"
        class="tm-form-msg--desc"
      />
    </TmFormGroup>
  </ActionModal>
</template>

<script>
import { mapState, mapGetters } from "vuex"
import { between, decimal } from "vuelidate/lib/validators"
import gql from "graphql-tag"
import { uatoms, SMALLEST } from "src/scripts/num"
import TmField from "src/components/common/TmField"
import TmFieldGroup from "src/components/common/TmFieldGroup"
import TmBtn from "src/components/common/TmBtn"
import TmFormGroup from "src/components/common/TmFormGroup"
import TmFormMsg from "src/components/common/TmFormMsg"
import ActionModal from "./ActionModal"
import transaction from "../utils/transactionTypes"
import { toMicroDenom } from "src/scripts/common"
import { formatBech32 } from "src/filters"
import { UserTransactionAdded } from "src/gql"

export default {
  name: `delegation-modal`,
  components: {
    TmField,
    TmFieldGroup,
    TmBtn,
    TmFormGroup,
    TmFormMsg,
    ActionModal
  },
  props: {
    targetValidator: {
      type: Object,
      required: true
    }
  },
  data: () => ({
    amount: null,
    selectedIndex: 0,
    balance: {
      amount: null,
      denom: ``
    },
    delegations: [],
    denom: ``
  }),
  computed: {
    ...mapState([`session`]),
    ...mapGetters([`network`, `address`]),
    fromOptions() {
      return [
        // from wallet
        {
          address: this.address,
          maximum: Number(this.balance.amount),
          key: `My Wallet - ${formatBech32(this.address, false, 20)}`,
          value: 0
        }
      ]
    },
    from() {
      if (!this.session.signedIn) return ``

      return this.fromOptions[this.selectedIndex].address
    },
    transactionData() {
      if (!this.targetValidator.operatorAddress) return {}

      return {
        type: transaction.DELEGATE,
        validatorAddress: this.targetValidator.operatorAddress,
        amount: uatoms(this.amount),
        denom: toMicroDenom(this.denom)
      }
    },
    notifyMessage() {
      return {
        title: `Successfully staked!`,
        body: `You have successfully staked your ${this.denom}s`
      }
    },
    maxAmount() {
      return this.fromOptions[this.selectedIndex].maximum
    }
  },
  mounted() {
    this.$apollo.queries.balance.refetch()
    this.$apollo.queries.delegations.refetch()
  },
  methods: {
    open(options) {
      if (options && options.redelegation && this.fromOptions.length > 1) {
        this.selectedIndex = 1
      }
      this.$refs.actionModal.open()
    },
    validateForm() {
      this.$v.$touch()

      return !this.$v.$invalid
    },
    clear() {
      this.$v.$reset()

      this.selectedIndex = 0
      this.amount = null
    },
    setMaxAmount() {
      this.amount = this.maxAmount
    },
    isMaxAmount() {
      return parseFloat(this.amount) === parseFloat(this.maxAmount)
    },
    enterPressed() {
      this.$refs.actionModal.validateChangeStep()
    },
    onSuccess(event) {
      this.$emit(`success`, event)
    }
  },
  validations() {
    return {
      amount: {
        required: x => !!x && x !== `0`,
        decimal,
        between: between(SMALLEST, this.maxAmount)
      }
    }
  },
  apollo: {
    delegations: {
      query: gql`
        query DelegationsDelegationModal(
          $networkId: String!
          $delegatorAddress: String!
        ) {
          delegations(
            networkId: $networkId
            delegatorAddress: $delegatorAddress
          ) {
            amount
            validator {
              operatorAddress
              name
            }
          }
        }
      `,
      skip() {
        return !this.address
      },
      variables() {
        return {
          networkId: this.network,
          delegatorAddress: this.address
        }
      },
      update(data) {
        /* istanbul ignore next */
        return data.delegations
      }
    },
    balance: {
      query: gql`
        query BalanceDelegationModal(
          $networkId: String!
          $address: String!
          $denom: String!
        ) {
          balance(networkId: $networkId, address: $address, denom: $denom) {
            amount
            denom
          }
        }
      `,
      skip() {
        return !this.address
      },
      variables() {
        return {
          networkId: this.network,
          address: this.address,
          denom: this.denom
        }
      }
    },
    denom: {
      query: gql`
        query NetworksDelegationModal($networkId: String!) {
          network(id: $networkId) {
            id
            stakingDenom
          }
        }
      `,
      variables() {
        return {
          networkId: this.network
        }
      },
      update(data) {
        /* istanbul ignore next */
        return data.network.stakingDenom
      }
    }
  },
  $subscribe: {
    userTransactionAdded: {
      variables() {
        return {
          networkId: this.network,
          address: this.address
        }
      },
      skip() {
        return !this.address
      },
      query: UserTransactionAdded,
      result({ data }) {
        if (data.userTransactionAdded.success) {
          this.$apollo.queries.delegations.refetch()
        }
      }
    }
  }
}
</script>
