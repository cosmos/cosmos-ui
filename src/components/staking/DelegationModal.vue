<template>
  <ActionModal
    id="delegation-modal"
    ref="actionModal"
    :submit-fn="submitForm"
    :simulate-fn="simulateForm"
    :validate="validateForm"
    :amount="amount"
    title="Delegate"
    class="delegation-modal"
    submission-error-prefix="Delegating failed"
    @close="clear"
  >
    <TmFormGroup class="action-modal-form-group" field-id="to" field-label="To">
      <TmField id="to" v-model="to" type="text" readonly />
    </TmFormGroup>

    <TmFormGroup
      v-if="fromOptions.length > 1"
      class="action-modal-form-group"
      field-id="from"
      field-label="From"
    >
      <TmField
        id="from"
        v-model="selectedIndex"
        :title="from"
        :options="fromOptions"
        type="select"
      />
    </TmFormGroup>
    <TmFormGroup
      :error="$v.amount.$error && $v.amount.$invalid"
      class="action-modal-form-group"
      field-id="amount"
      field-label="Amount"
    >
      <span class="input-suffix">{{ num.viewDenom(denom) }}</span>
      <TmField
        id="amount"
        v-model="amount"
        type="number"
        placeholder="Amount"
      />
      <TmFormMsg
        v-if="balance === 0"
        :msg="`doesn't have any ${num.viewDenom(denom)}s`"
        name="Wallet"
        type="custom"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && !$v.amount.decimal"
        name="Amount"
        type="numeric"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && (!$v.amount.required || amount === 0)"
        name="Amount"
        type="required"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && !$v.amount.between"
        :max="$v.amount.$params.between.max"
        :min="$v.amount.$params.between.min"
        name="Amount"
        type="between"
      />
    </TmFormGroup>
  </ActionModal>
</template>

<script>
import { mapGetters } from "vuex"
import { between, decimal } from "vuelidate/lib/validators"
import num, { uatoms, atoms, SMALLEST } from "../../scripts/num.js"
import TmField from "common/TmField"
import TmFormGroup from "common/TmFormGroup"
import TmFormMsg from "common/TmFormMsg"
import ActionModal from "common/ActionModal"

export default {
  name: `delegation-modal`,
  components: {
    TmField,
    TmFormGroup,
    TmFormMsg,
    ActionModal
  },
  props: {
    fromOptions: {
      type: Array,
      required: true
    },
    to: {
      type: String,
      required: true
    },
    validator: {
      type: Object,
      required: true
    },
    denom: {
      type: String,
      required: true
    }
  },
  data: () => ({
    amount: null,
    selectedIndex: 0,
    num
  }),
  computed: {
    ...mapGetters([`delegates`, `session`, `bondDenom`]),
    balance() {
      if (!this.session.signedIn) return 0

      return this.fromOptions[this.selectedIndex].maximum
    },
    from() {
      if (!this.session.signedIn) return ``

      return this.fromOptions[this.selectedIndex].address
    }
  },
  methods: {
    open() {
      this.$refs.actionModal.open()
    },
    validateForm() {
      this.$v.$touch()

      return !this.$v.$invalid
    },
    clear() {
      this.$v.$reset()

      this.selectedIndex = 0
      this.amount = null
    },
    async simulateDelegation() {
      return await this.$store.dispatch(`simulateDelegation`, {
        validator_address: this.validator.operator_address,
        amount: String(uatoms(this.amount))
      })
    },
    async submitDelegation(gasEstimate, gasPrice, password, submitType) {
      await this.$store.dispatch(`submitDelegation`, {
        validator_address: this.validator.operator_address,
        amount: String(uatoms(this.amount)),
        submitType,
        password,
        gas: String(gasEstimate),
        gas_prices: [
          {
            amount: String(uatoms(gasPrice)),
            denom: this.denom
          }
        ]
      })

      this.$store.commit(`notify`, {
        title: `Successful delegation!`,
        body: `You have successfully delegated your ${num.viewDenom(
          this.denom
        )}s`
      })
    },
    async simulateRedelegation() {
      const validatorSrc = this.delegates.delegates.find(
        v => this.from === v.operator_address
      )
      return await this.$store.dispatch(`simulateRedelegation`, {
        validatorSrc,
        validatorDst: this.validator,
        amount: String(uatoms(this.amount))
      })
    },
    async submitRedelegation(gasEstimate, gasPrice, password, submitType) {
      const validatorSrc = this.delegates.delegates.find(
        v => this.from === v.operator_address
      )
      await this.$store.dispatch(`submitRedelegation`, {
        validatorSrc,
        validatorDst: this.validator,
        amount: {
          amount: String(uatoms(this.amount)),
          denom: this.denom
        },
        submitType,
        password,
        gas: String(gasEstimate),
        gas_prices: [
          {
            amount: String(uatoms(gasPrice)),
            denom: this.denom
          }
        ]
      })

      this.$store.commit(`notify`, {
        title: `Successful redelegation!`,
        body: `You have successfully redelegated your ${num.viewDenom(
          this.denom
        )}s`
      })
    },
    async simulateForm() {
      if (this.from === this.session.address) {
        return await this.simulateDelegation()
      } else {
        return await this.simulateRedelegation()
      }
    },
    async submitForm(gasEstimate, gasPrice, password, submitType) {
      if (this.from === this.session.address) {
        await this.submitDelegation(gasEstimate, gasPrice, password, submitType)
      } else {
        await this.submitRedelegation(
          gasEstimate,
          gasPrice,
          password,
          submitType
        )
      }
    }
  },
  validations() {
    return {
      amount: {
        required: x => !!x && x !== `0`,
        decimal,
        between: between(SMALLEST, atoms(this.balance))
      }
    }
  }
}
</script>
