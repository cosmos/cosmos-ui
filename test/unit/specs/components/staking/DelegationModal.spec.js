"use strict"

import htmlBeautify from "html-beautify"
import { createLocalVue, mount } from "@vue/test-utils"
import DelegationModal from "staking/DelegationModal"
import Vuelidate from "vuelidate"

const Wrapper = () => {
  const localVue = createLocalVue()
  localVue.use(Vuelidate)

  return mount(DelegationModal, {
    localVue,
    propsData: {
      fromOptions: [
        {
          address: `cosmosaccaddr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctpesxxn9`,
          key: `My Wallet - cosmosaccaddr…3p4mqpjyrm5ctpesxxn9`,
          maximum: 100,
          value: 0
        },
        {
          address: `cosmosaccaddr126ayk3hse5zvk9gxfmpsjr9565ef72pv9g20yx`,
          key: `Billy the Bill - cosmosaccaddr…psjr9565ef72pv9g20yx`,
          maximum: 23.0484375481,
          value: 1
        },
        {
          address: `cosmosaccaddr18thamkhnj9wz8pa4nhnp9rldprgant57ryzag7`,
          key: `Kentucky - cosmosaccaddr…np9rldprgant57ryzag7`,
          maximum: 1.3788878447,
          value: 2
        }
      ],
      to: `cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctplpn3au`,
      bondingDenom: `Stake`
    }
  })
}

test(`renders correctly`, () => {
  expect(htmlBeautify(Wrapper().html())).toMatchSnapshot()
})

test(`the 'amount' field defaults to 0`, () => {
  let amountField = Wrapper().find(`#amount`)
  expect(Number(amountField.element.value)).toEqual(0)
})

test(`display the 'To' address`, () => {
  let toField = Wrapper().find(`#to`)
  expect(toField.element.value).toEqual(
    `cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctplpn3au`
  )
})

test(`Delegation button submits delegation and closes modal`, () => {
  const wrapper = Wrapper()
  wrapper.setData({ amount: 50 })
  wrapper.vm.onDelegation()

  expect(wrapper.emittedByOrder()).toEqual([
    {
      name: `submitDelegation`,
      args: [
        {
          amount: 50,
          from: `cosmosaccaddr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctpesxxn9`
        }
      ]
    },
    {
      name: `update:showDelegationModal`,
      args: [false]
    }
  ])
})

test(`X button emits close signal`, () => {
  const wrapper = Wrapper()
  wrapper.vm.close()

  expect(wrapper.emittedByOrder()).toEqual([
    {
      name: `update:showDelegationModal`,
      args: [false]
    }
  ])
})
