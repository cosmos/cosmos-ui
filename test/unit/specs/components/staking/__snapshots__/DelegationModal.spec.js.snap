// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
"<div id=\\"delegation-modal\\" class=\\"delegation-modal\\">
   <div class=\\"delegation-modal-header\\">
      <img src=\\"~assets/images/cosmos-logo.png\\" class=\\"icon delegation-modal-atom\\"><span class=\\"tm-modal-title\\">Delegation</span>
      <div id=\\"closeBtn\\" class=\\"tm-modal-icon tm-modal-close\\">
         <i class=\\"material-icons\\">close</i>
      </div>
</div>
<div class=\\"delegation-modal-form-group tm-form-group\\">
   <!---->
   <label for=\\"amount\\" class=\\"tm-form-group__label\\">Amount</label>
   <div class=\\"tm-form-group__field\\">
      <input type=\\"number\\" placeholder=\\"atom\\" class=\\"tm-field\\" id=\\"denom\\" readonly=\\"readonly\\"><input type=\\"number\\" max=\\"100\\" min=\\"0\\" class=\\"tm-field\\" id=\\"amount\\" step=\\"1\\">
   </div>
</div>
<div class=\\"delegation-modal-form-group tm-form-group\\">
   <!---->
   <label for=\\"to\\" class=\\"tm-form-group__label\\">To</label>
   <div class=\\"tm-form-group__field\\">
      <input class=\\"tm-field\\" id=\\"to\\" readonly=\\"readonly\\">
   </div>
</div>
<div class=\\"delegation-modal-form-group tm-form-group\\">
   <!---->
   <label for=\\"from\\" class=\\"tm-form-group__label\\">From</label>
   <div class=\\"tm-form-group__field\\">
      <div class=\\"tm-select\\" id=\\"from\\" title=\\"cosmos15ky9du8a2wlstz6fpx3p4mqpjyrm5ctpesxxn9\\">
         <select class=\\"tm-field tm-field-select\\">
         <option value=\\"\\" disabled=\\"disabled\\" selected=\\"selected\\" hidden=\\"hidden\\">
            Select option...
         </option>
         <option value=\\"0\\">
            My Wallet - cosmos…3p4mqpjyrm5ctpesxxn9
         </option>
         <option value=\\"1\\">
            Billy the Bill - cosmos…psjr9565ef72pv9g20yx
         </option>
         <option value=\\"2\\">
            Kentucky - cosmos…np9rldprgant57ryzag7
         </option>
         </select>
         <div class=\\"tm-field-select-addon\\">
            <i class=\\"material-icons\\">arrow_drop_down</i>
         </div>
</div>
</div>
</div>
<div class=\\"delegation-modal-footer\\">
   <button class=\\"tm-btn\\" disabled=\\"disabled\\"><span class=\\"tm-btn__container tm-btn--size-lg tm-btn--primary\\">
   <!---->
   <!---->
   <span class=\\"tm-btn__value\\">Confirm Delegation</span></span></button>
</div>
</div>"
`;
